File: core/src/main/scala/com/electronwill/collections/ConcurrentBag.java
package com.electronwill.collections;

import java.util.Arrays;
import java.util.Iterator;

/**
 * A thread-safe Bag of elements.
 *
 * @author TheElectronWill
 */
public final class ConcurrentBag<E> extends AbstractBag<E> implements Bag<E> {
	private volatile Object[] array;
	private volatile int size;
	private final int capacityIncrement;

	public ConcurrentBag() {
		this(10, 50);
	}

	public ConcurrentBag(int initialCapacity) {
		this(initialCapacity, 50);
	}

	public ConcurrentBag(int initialCapacity, int capacityIncrement) {
		this.array = new Object[initialCapacity];
		this.capacityIncrement = capacityIncrement;
	}

	@Override
	public E get(int index) {
		final Object[] arr = array;// volatile read
		return (E)arr[index];
	}

	@Override
	public E tryGet(int index) {
		final Object[] arr = array;// volatile read
		return (index >= 0 && index < arr.length) ? (E)arr[index] : null;
	}

	@Override
	public E remove(int index) {
		synchronized (this) {
			Object[] arr = array;
			if (index < 0 || index >= arr.length) {
				return null;
			}
			E element = (E)arr[index];
			if (element != null) {
				int s = size;
				arr[index] = arr[--s];
				arr[s] = null;
				size = s;
			}
			return element;
		}
	}

	@Override
	public void compact() {
		synchronized (this) {
			array = Arrays.copyOf(array, size);
		}
	}

	@Override
	public int size() {
		return size;
	}

	@Override
	public boolean isEmpty() {
		return size == 0;
	}

	@Override
	public boolean contains(Object o) {
		Object[] arr = array;
		int s = Math.min(size, arr.length);
		for (int i = 0; i < s; i++) {
			Object e = arr[i];
			if (e.equals(o)) {
				return true;
			}
		}
		return false;
	}

	@Override
	public Iterator<E> iterator() {
		return new ConcurrentBagIterator();
	}

	@Override
	public Object[] toArray() {
		final Object[] arr = array;
		return Arrays.copyOf(arr, Math.min(size, arr.length));
		//The array could have grown between the volatile get and copyOf
	}

	@Override
	public boolean add(E e) {
		synchronized (this) {
			if (size == array.length) {
				array = Arrays.copyOf(array, size + capacityIncrement);
			}
			array[size++] = e;
			return true;
		}
	}

	@Override
	public boolean remove(Object o) {
		synchronized (this) {
			final int s = size;
			final Object[] arr = array;
			for (int i = 0; i < s; i++) {
				if (arr[i].equals(o)) {
					remove(i);
					return true;
				}
			}
			return false;
		}
	}

	@Override
	public void clear() {
		synchronized (this) {
			array = new Object[10];
			size = 0;
		}
	}

	/**
	 * A weakly consistent entry iterator. It is not guaranteed to reflect the latest state of the
	 * bag but will never throw ConcurrentModificationException.
	 */
	private class ConcurrentBagIterator implements Iterator<E> {
		private final Object[] array;//The array at the time the iterator is created
		private int pos = 0;//The next position
		private E returnedElement;//The last element returned by next

		private ConcurrentBagIterator() {
			this.array = ConcurrentBag.this.array;
		}

		@Override
		public boolean hasNext() {
			return pos < Math.min(size, array.length);
		}

		@Override
		public E next() {
			returnedElement = (E)array[pos++];
			return returnedElement;
		}

		@Override
		public void remove() {
			if (returnedElement == null) {
				throw new IllegalStateException("remove() already called for this element");
			}
			synchronized (ConcurrentBag.this) {
				if (array != ConcurrentBag.this.array) {// The array has changed
					ConcurrentBag.this.remove(returnedElement);// Removes by object
				} else {
					ConcurrentBag.this.remove(--pos);// Removes by index
				}
			}
			returnedElement = null;
		}
	}
}

File: utils/src/com/electronwill/collection/ConcurrentBag.scala
package com.electronwill.collection

import scala.reflect.ClassTag

/**
 * Thread-safe implementation of [[Bag]] whose [[apply]] method doesn't use `synchronized` nor a lock.
 *
 * @author TheElectronWill
 */
final class ConcurrentBag[A >: Null <: AnyRef: ClassTag](initialCapacity: Int = 16) extends Bag[A] {
  /** Current bag size */
  @volatile private[this] var s = 0

  /** Elements storage */
  @volatile private[this] var array = new Array[A](initialCapacity)

  override def apply(i: Int): A = {
    if (i >= 0 && i < s) {
      array(i)
    } else {
      throw new IndexOutOfBoundsException("Invalid index: " + i)
    }
  }
  override def size: Int = s

  override def remove(i: Int): Unit = {
    this.synchronized {
      val length = s
      if (i >= 0 && i < length) {
        val updatedArray = array
        val newLength = length - 1
        updatedArray(i) = updatedArray(newLength)
        updatedArray(newLength) = null
        array = updatedArray
        s = newLength
      }
    }
  }
  override def +=(elem: A): ConcurrentBag.this.type = {
    this.synchronized {
      val length = s
      var updatedArray = array
      if (length == updatedArray.length) {
        updatedArray = grow(updatedArray, length + (length >> 1))
      }
      updatedArray(length) = elem
      array = updatedArray
      s = length + 1
    }
    this
  }
  override def ++=(arr: Array[A], offset: Int, length: Int): this.type = {
    this.synchronized {
      val newS = s + length
      if (newS >= array.length) {
        array = grow(array, math.max(newS, s + (s >> 1)))
      }
      System.arraycopy(arr, offset, array, s, length)
    }
    this
  }
  override def -=(elem: A): ConcurrentBag.this.type = {
    this.synchronized {
      val length = s
      val upToDateArray = array
      var i = 0
      while (i < length && upToDateArray(i) != elem) {
        i += 1
      }
      remove(i)
    }
    this
  }
  override def clear(): Unit = {
    this.synchronized {
      val length = s
      val updatedArray = array
      var i = 0
      while (i < length) {
        updatedArray(i) = null
        i += 1
      }
      array = updatedArray
      s = 0
    }
  }
  override def compact(): Unit = {
    if (s < array.length) {
      this.synchronized {
        array = shrink(array, s)
      }
    }
  }
  override def indexOf(elem: A): Int = {
    this.synchronized {
      var i = 0
      while (i < s && array(i) != elem) {
        i += 1
      }
      if (i == s) -1 else i
    }
  }
  override def iterator: MutableIterator[A] = new MutableIterator[A] {
    private[this] var i = 0
    private[this] val (arrayView: Array[A], sView: Int) = {
      var av: Array[A] = null
      var sv = 0
      do {
        av = array
        sv = s
      } while (array ne av)
      (av, sv)
    }
    override def hasNext: Boolean = {
      i < sView
    }
    override def next(): A = {
      val elem = arrayView(i)
      i += 1
      elem
    }
    override def remove(): Unit = {
      ConcurrentBag.this.remove(i)
    }
    override def insert(elem: A): Unit = {
      ConcurrentBag.this.+=(elem)
    }
  }
}
